File: app/src/main/java/com/chiron/playpalyer/MainActivity.java
package com.chiron.playpalyer;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.media.MediaCodecInfo;
import android.media.MediaCodecList;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.chiron.playpalyer.utils.PermissionUtil;

public class MainActivity extends AppCompatActivity implements View.OnClickListener{
    private static final int REQUEST_CODE = 1;
    private static final String TAG = MainActivity.class.getSimpleName();
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        PermissionUtil permissionUtil = new PermissionUtil();
        permissionUtil.verifyPermission(this);
        initUi();

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            MediaCodecList list = new MediaCodecList(MediaCodecList.REGULAR_CODECS);
            MediaCodecInfo[] supportCodes = list.getCodecInfos();
            Log.i(TAG, "解码器列表：");
            for (MediaCodecInfo codec : supportCodes) {
                if (!codec.isEncoder()) {
                    String name = codec.getName();
                    if (name.startsWith("OMX.google")) {
                        Log.i(TAG, "软解->" + name);
                    }
                }
            }
            for (MediaCodecInfo codec : supportCodes) {
                if (!codec.isEncoder()) {
                    String name = codec.getName();
                    if (!name.startsWith("OMX.google")) {
                        Log.i(TAG, "硬解->" + name);
                    }
                }
            }
            Log.i(TAG, "编码器列表：");
            for (MediaCodecInfo codec : supportCodes) {
                if (codec.isEncoder()) {
                    String name = codec.getName();
                    if (name.startsWith("OMX.google")) {
                        Log.i(TAG, "软编->" + name);
                    }
                }
            }
            for (MediaCodecInfo codec : supportCodes) {
                if (codec.isEncoder()) {
                    String name = codec.getName();
                    if (!name.startsWith("OMX.google")) {
                        Log.i(TAG, "硬编->" + name);
                    }
                }
            }

            Log.i(TAG,"硬件=>"+Build.HARDWARE);
        }
    }

    private void initUi(){
        findViewById(R.id.btn_media).setOnClickListener(this);
        findViewById(R.id.btn_recorder).setOnClickListener(this);
        findViewById(R.id.btn_new_recorder).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_media: {
                Intent intent = new Intent(MainActivity.this, MediaActivity.class);
                startActivity(intent);
                break;
            }

            case R.id.btn_recorder: {
                Intent intent = new Intent(MainActivity.this, RecorderActivity.class);
                startActivity(intent);
                break;
            }

            case R.id.btn_new_recorder:{
                Intent intent = new Intent(MainActivity.this, NewRecorderActivity.class);
                startActivity(intent);
                break;
            }
            default:
                break;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode){
            case REQUEST_CODE:
                if(grantResults.length>0 && grantResults[0]== PackageManager.PERMISSION_GRANTED){
                    Toast.makeText(this,"权限申请通过",Toast.LENGTH_SHORT);
                }else{
                    Toast.makeText(this,"权限申请失败",Toast.LENGTH_SHORT);
                }
                break;
            default:
                super.onRequestPermissionsResult(requestCode,permissions,grantResults);
        }
    }
}

File: app/src/main/java/com/chiron/playpalyer/MainActivity.kt
package com.chiron.playpalyer

import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Environment
import android.view.View
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.chiron.playpalyer.mediac.decoder.AudioDecoder
import com.chiron.playpalyer.mediac.decoder.VideoDecoder
import com.chiron.playpalyer.mediac.muxer.MP4Repack
import com.chiron.playpalyer.permission.Permission
import kotlinx.android.synthetic.main.activity_main.*
import java.util.concurrent.Executors
import java.util.jar.Manifest

class MainActivity : AppCompatActivity() {
    val path=Environment.getExternalStorageDirectory().absolutePath+"/mvtest.mp4"
    lateinit var videoDecoder: VideoDecoder
    lateinit var audioDecoder: AudioDecoder
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val permission= Permission()
        permission.verifyStoragePermission(this)
        initPlayer()
    }

    private fun initPlayer(){
        val threadPool= Executors.newFixedThreadPool(2)
        videoDecoder=VideoDecoder(path,sfv,null)
        threadPool.execute(videoDecoder)

        audioDecoder=AudioDecoder(path)
        threadPool.execute(audioDecoder)

        videoDecoder.resume()
        audioDecoder.resume()
    }

    fun clickRepack(view:View){
        repack()
    }

    private fun repack(){
        //视频输入路径
        val path=Environment.getExternalStorageDirectory().absolutePath+"/mvtest.mp4"
        val repack=MP4Repack(path)
        repack.start()
    }

    override fun onDestroy() {
        videoDecoder.stop()
        audioDecoder.stop()
        super.onDestroy()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        when (requestCode){
            1->{
                if(grantResults.size>0 && grantResults[0]==PackageManager.PERMISSION_GRANTED){
                    Toast.makeText(this,"权限申请通过",Toast.LENGTH_LONG)
                }else{
                    Toast.makeText(this,"权限申请不通过",Toast.LENGTH_LONG)
                    finish()
                }
            }
            else->{
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
            }
        }

    }


}